### PR TITLE
Update to kzs_attachement

### PR DESCRIPTION
**Pull request by tsromana**
added use case where users att creds could be changed and at that point amazon s3 will disregard old creds and will use newer ones instead. 

While sending a request to get doc, it will check if the docs att handler and the systems att handler are same if not it will use the docs att handler creds, 
if they are same it will use the systems creds and if somehow the creds are messed up it will revert back to att's creds.

Todo:
Needs thorough testing

### CODE REVIEW

**Comment by lazedo** [comment]
@tsromana thanks. i'll take a look at this next week

**Comment by tsromana** [comment]
no problem I also didn't get time yet to test this properly :) hopefully
will add some changes if I find some bug in it before your review

**Comment by lazedo** [comment]
@tsromana LGTM, can you just rename `system_props` to `handler_props` ?

**Comment by tsromana** [comment]
@lazedo done!

File: core/kazoo_attachments/src/kz_att_s3.erl [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
 -spec put_attachment(kz_data:connection(), ne_binary(), ne_binary(), ne_binary(), ne_binary(), kz_data:options()) -> any().
 put_attachment(Params, DbName, DocId, AName, Contents, _Options) ->
-    {Bucket, Key, Secret, Path, Host} = get_map_values(Params),
+    {Key, Secret, Host, Bucket, Path} = get_map_values(Params),
```
Comment (lazedo) [inline core/kazoo_attachments/src/kz_att_s3.erl]: @tsromana this is an unrelated change

File: core/kazoo_attachments/src/kz_att_s3.erl [hunk 2]
```diff
@@ -63,26 +64,36 @@ convert_kv({<<"etag">> = K, V}) ->
...
+-spec get_map_values(kz_data:connection()) ->
+                            {ne_binary(), ne_binary(), ne_binary(), ne_binary(), api_binary()}.
 get_map_values(#{'bucket' := Bucket
                 ,'key' := Key
                 ,'secret' := Secret
                 }=Map) ->
     Path = maps:get('path', Map, 'undefined'),
     Host = maps:get('host', Map,  ?AMAZON_S3_HOST),
-    {Bucket, Key, Secret, Path, Host}.
+    {Key, Secret, Host, Bucket, Path}.
```
Comment (lazedo) [inline core/kazoo_attachments/src/kz_att_s3.erl]: unrelated change

File: core/kazoo_attachments/src/kz_att_s3.erl [hunk 3]
```diff
@@ -63,26 +64,36 @@ convert_kv({<<"etag">> = K, V}) ->
...
 get_file_path(Path, DbName, DocId, AName) ->
     kz_util:to_list(list_to_binary([Path, "/", DbName, "/", DocId, "_", AName])).
 
--spec get_s3_values(ne_binary()) -> {ne_binary(), ne_binary(), ne_binary(), ne_binary(), api_binary()}.
-get_s3_values(S3) ->
+-spec get_s3_values(ne_binary(), kz_data:connection()) ->
+                           {ne_binary(), ne_binary(), ne_binary(), ne_binary(), api_binary()}.
+get_s3_values(S3, 'undefined') ->
     case binary_to_term(base64:decode(S3)) of
         {Key, Secret, Bucket, Path} ->
```
Comment (lazedo) [inline core/kazoo_attachments/src/kz_att_s3.erl]: i think this should be fixed  to {Bucket, Key, Secret, Path}

Comment (tsromana) [inline core/kazoo_attachments/src/kz_att_s3.erl]: @lazedo reverted back to same values, sorry was trying to aim for same output on both funs :)

Comment (tsromana) [inline core/kazoo_attachments/src/kz_att_s3.erl]: @lazedo these values are the default which we had earlier

File: core/kazoo_attachments/src/kz_att_s3.erl [hunk 4]
```diff
@@ -78,11 +80,21 @@ get_file_path('undefined', DbName, DocId, AName) ->
...
             {Key, Secret, ?AMAZON_S3_HOST, Bucket, Path};
         {_Key, _Secret, _Host, _Bucket, _Path} = V2 ->
             V2
+    end;
+get_s3_values(S3, HandlerProps) ->
+    try get_map_values(HandlerProps) of
+        {Bucket, Key, Secret, Path, Host} ->
+            {Key, Secret, Host, Bucket, Path}
```
Comment (lazedo) [inline core/kazoo_attachments/src/kz_att_s3.erl]: @tsromana we should preserve initial `Host, Bucket, Path` values

Comment (tsromana) [inline core/kazoo_attachments/src/kz_att_s3.erl]: @lazedo i have changed the values, let me know if the var naming is correct or should i change it something else :)

**Comment by tsromana** [comment]
@lazedo just to clarify the att handler we have for google is for google drive not for google cloud storage right

**Comment by lazedo** [comment]
@tsromana looks good. circleci is failing. can you try to rebase to latest master ? tks

**Comment by tsromana** [comment]
@lazedo will do :)

**Comment by lazedo** [comment]
@tsromana looks good but, still reflecting on a possible issue.
if we save to Host = A with Secret = Y
when we fetch, we may be using Secret = Z configured for Host = B.
i think we should only override the `secret` if the host is the same.
thoughts ?

**Comment by lazedo** [comment]
@tsromana so after IRC discussion, the new `key`and `secret` should only be used if `Host` and `bucket` from handler are equal to what we have in S3 object.
thanks

**Comment by tsromana** [comment]
@lazedo will do and use the old ones if not same :)

File: core/kazoo_attachments/src/kz_att_s3.erl [hunk 5]
```diff
@@ -78,11 +80,24 @@ get_file_path('undefined', DbName, DocId, AName) ->
...
+    {_Key, _Secret, Host, Bucket, Path} = get_s3_values(S3, 'undefined'),
+    try get_map_values(HandlerProps) of
+        {Bucket, Key, Secret, _Path, Host} ->
+            {Key, Secret, Host, Bucket, Path};
+        {_Bucket, Key, Secret, _Path, _Host} ->
+            {_Key, _Secret, Host, Bucket, Path}
+    catch
+        _E:_R ->
+            lager:info("unable to get s3 cred value using att values (~s): ~p", [_E, _R]),
+            {_Key, _Secret, Host, Bucket, Path}
```
Comment (lazedo) [inline core/kazoo_attachments/src/kz_att_s3.erl]: @tsromana we usualy use `underscore` for variables that we really don't use besides logging
how about
```
    Original =  {_, _, Host, Bucket, Path} = get_s3_values(S3, 'undefined'),
    try get_map_values(HandlerProps) of
        {Bucket, Key, Secret, _Path, Host} ->
            {Key, Secret, Host, Bucket, Path};
        _ -> Original
    catch
...
```

Comment (tsromana) [inline core/kazoo_attachments/src/kz_att_s3.erl]: @lazedo yea this looks more readable :)
